File: dijkstra_shortest_path/dijkstra.py
"""Dijkstra's shortest path algorithm"""
from hash_heap import HashHeap
from graph import Graph
# import ipdb; ipdb.set_trace()


def dijkstra(graph, source):
    """Dijkstra

    This method computes shortest distance and path for each node
    from the source node, and save results in each node's distance
    attribute.
    return None
    """
    reachable = preprocessing(graph, source)
    for id in reachable:
        graph[id].dijkstra_dist = float('inf')
    source.dijkstra_dist = 0
    unknown = HashHeap(key=lambda x, y: x.dijkstra_dist < y.dijkstra_dist)
    for node_id in reachable:
        unknown.add(graph[node_id])  # add to unknown
    visited = set()

    while unknown:
        node = unknown.pop()
        visited.add(node.id)  # add to known
        for edge in node.edges:
            adj = graph[edge.dest]
            if (edge.dest not in visited and
                    node.dijkstra_dist + edge.dist < adj.dijkstra_dist):
                adj.dijkstra_dist = node.dijkstra_dist + edge.dist
                unknown.update(adj)  # method only in hash heap

def preprocessing(graph, source):
    """
    Return reachable nodes id as a list
    """
    marked = [False] * graph.n
    dfs(graph, source, marked)
    return [i for i, item in enumerate(marked) if item]

def dfs(graph, node, marked):
    marked[node.id] = True
    for edge in node.edges:
        adj = graph[edge.dest]
        if not marked[adj.id]:
            dfs(graph, adj, marked)

File: graphs/all_pairs_shortest_paths/Dijkstra/dijkstra.cpp
#include "dijkstra.h"

int bfs(int sID, Graph g);
std::vector<Graph::Dist> dijkstra(int sID, Graph g)
// compute the shortest paths lengths from s to each node in g
// input: source node id sID, graph g
// output: a list of distances from s to each node in graph g
{
    Node &s = g.graph[sID];
    s.dist.n = 0; s.dist.inf = false;
    int knownSize = 0;
    auto cmp = [](const Node &n1, const Node &n2) {
        return n1.dist.n > n2.dist.n;  // min-heap
    };
    std::priority_queue<Node, std::vector<Node>, decltype(cmp)> unknown(cmp);
    unknown.push(s);
    int numReachableNodes = bfs(sID, g);
    while (knownSize < numReachableNodes) {
        int nID = unknown.top().nodeID; unknown.pop();
        Node &n = g.graph[nID];
        if (!n.known) {
            n.known = true;
            ++knownSize;
            for (Edge &e : n.edges) {
                Node &adj = g.graph[e.to];
                if (!adj.known) {
                    if (adj.dist.inf || adj.dist.n > n.dist.n + e.cost) {
                        adj.dist.inf = false;
                        adj.dist.n = n.dist.n + e.cost;
                        unknown.push(adj);
                    }
                }
            }
        }
    }
    std::vector<Dist> ret;
    for (Node &n : g.graph) {
        ret.push_back(n.dist);
    }
    return ret;
}

int bfs(int sID, Graph g)
// bfs O(m)
// input: source node id sID, graph g
// output: number of nodes that are reachable from sID, including sID itself
{
    Node &n = g.graph[sID];
    n.known = true;
    int numReadchableNodes = 1;
    std::queue<int> q;
    q.push(sID);
    while (!q.empty()) {
        int nID = q.front(); q.pop();
        Node &n = g.graph[nID];
        for (Edge &e : n.edges) {
            int adjID = e.to;
            Node &adjNode = g.graph[adjID];
            if (!adjNode.known) {
                adjNode.known = true;
                q.push(adjID);
                ++numReadchableNodes;
            }
        }
    }
    return numReadchableNodes;
}